Snowflake.Data.Tests/IntegrationTests/StructuredTypesIT.cs

```diff
@@ -0,0 +1,1471 @@
+                    Assert.IsNull(allTypesObject.FloatValue);
+                    Assert.IsNull(allTypesObject.DoubleValue);
+                    Assert.IsNull(allTypesObject.DecimalValue);
+                    Assert.IsNull(allTypesObject.BooleanValue);
+                    Assert.IsNull(allTypesObject.GuidValue);
+                    Assert.IsNull(allTypesObject.DateTimeValue);
+                    Assert.IsNull(allTypesObject.DateTimeOffsetValue);
+                    Assert.IsNull(allTypesObject.TimeSpanValue);
+                    Assert.IsNull(allTypesObject.BinaryValue);
+                    Assert.IsNull(allTypesObject.SemiStructuredValue);
+                }
+            }
+        }
+
+        [Test]
+        public void TestSelectMapSkippingNullValues()
+        {
+            // arrange
+            using (var connection = new SnowflakeDbConnection(ConnectionString))
+            {
+                connection.Open();
+                using (var command = connection.CreateCommand())
+                {
+                    EnableStructuredTypes(connection);
+                    var mapAsSFString = "OBJECT_CONSTRUCT('a', NULL, 'b', '3')::MAP(VARCHAR, INTEGER)";
+                    command.CommandText = $"SELECT {mapAsSFString}";
+                    var reader = (SnowflakeDbDataReader)command.ExecuteReader();
+                    Assert.IsTrue(reader.Read());
+
+                    // act
+                    var map = reader.GetMap<string, int?>(0);
+
+                    // assert
+                    Assert.AreEqual(1, map.Count);
+                    Assert.AreEqual(3, map["b"]);
+                }
+            }
+        }
+
+        [Test]
+        [TestCase(@"OBJECT_CONSTRUCT('Value', OBJECT_CONSTRUCT('a', 'b'))::OBJECT(Value OBJECT)", "{\n  \"a\": \"b\"\n}")]
+        [TestCase(@"OBJECT_CONSTRUCT('Value', ARRAY_CONSTRUCT('a', 'b'))::OBJECT(Value ARRAY)", "[\n  \"a\",\n  \"b\"\n]")]
+        [TestCase(@"OBJECT_CONSTRUCT('Value', TO_VARIANT(OBJECT_CONSTRUCT('a', 'b')))::OBJECT(Value VARIANT)", "{\n  \"a\": \"b\"\n}")]
+        public void TestSelectSemiStructuredTypesInObject(string valueSfString, string expectedValue)
+        {
+            // arrange
+            using (var connection = new SnowflakeDbConnection(ConnectionString))
+            {
+                connection.Open();
+                using (var command = connection.CreateCommand())
+                {
+                    EnableStructuredTypes(connection);
+                    command.CommandText = $"SELECT {valueSfString}";
+                    var reader = (SnowflakeDbDataReader) command.ExecuteReader();
+                    Assert.IsTrue(reader.Read());
+
+                    // act
+                    var wrapperObject = reader.GetObject<StringWrapper>(0);
+
+                    // assert
+                    Assert.NotNull(wrapperObject);
+                    Assert.AreEqual(ConvertNewlinesOnWindows(expectedValue), wrapperObject.Value);
```

Comment:
I'm not sure this is the best test. What if backend decides upon changing two spaces to four? Or to tabs? It's better to skip comparing whitespaces.

Reply:
ok, I changed to removing whitespaces